gptqmodel/utils/threadx.py

```diff
@@ -75,19 +75,95 @@ def _coerce_device(d: DeviceLike) -> torch.device:
+                                torch.cuda.set_device(0)
+                        else:
+                            # Try to extract device index from string
+                            try:
+                                idx = int(target.split(":")[1]) if ":" in target else 0
+                                torch.cuda.set_device(idx)
+                            except (ValueError, IndexError):
+                                if torch.cuda.is_available():
+                                    torch.cuda.set_device(0)
+                    yield
+                finally:
+                    # Restore previous device
+                    if prev_device is not None:
+                        torch.cuda.set_device(prev_device)
+            else:
+                raise
     elif dev.type == "xpu" and hasattr(torch, "xpu"):
-        target = dev if dev.index is not None else "xpu"
-        with torch.xpu.device(target):
-            yield
+        # Fix: Use device index or string, not torch.device object directly
+        target = dev.index if dev.index is not None else "xpu"
+        if DEBUG_ON:
+            log.debug(f"_device_ctx: XPU device={dev}, target={target}, target_type={type(target)}")
+        
+        try:
+            with torch.xpu.device(target):
+                yield
+        except AttributeError as e:
+            if "prev_idx" in str(e):
+                log.error(f"_device_ctx: PyTorch version incompatibility - {e}")
+                log.error(f"_device_ctx: Falling back to direct device setting without context manager")
+                
+                # Capture previous device
+                prev_device = torch.xpu.current_device() if hasattr(torch.xpu, 'current_device') else None
+
+                # Fallback for XPU
+                try:
+                    if isinstance(target, int):
+                        torch.xpu.set_device(target)
+                    elif isinstance(target, str):
+                        if target == "xpu":
+                            if hasattr(torch.xpu, 'is_available') and torch.xpu.is_available():
+                                torch.xpu.set_device(0)
+                        else:
+                            try:
+                                idx = int(target.split(":")[1]) if ":" in target else 0
+                                torch.xpu.set_device(idx)
+                            except (ValueError, IndexError):
+                                if hasattr(torch.xpu, 'is_available') and torch.xpu.is_available():
+                                    torch.xpu.set_device(0)
+                    yield
+                finally:
+                    if prev_device is not None and hasattr(torch.xpu, 'set_device'):
+                        torch.xpu.set_device(prev_device)
+            else:
+                raise
     else:
+        if DEBUG_ON:
+            log.debug(f"_device_ctx: Non-CUDA/XPU device={dev}, yielding without context")
         yield
```

Comment:
I am confused at this block of code change. 

What is `prev_idx` and why are we using it?

Reply:
For me it randomly rarely failing with error, so that was a fix, I can revert and try if it still actual

Reply:
Are you testing with intel `xpu`? Let me know if it crashes again and record the stack trace. THe code appears to hack around some  `xpu` context bug which that triggered your crash. If it is internal to xpu code, I can forward the issue to the Intel team.

Reply:
no it was crashing with cuda - but the code is same for all device type

Reply:
no errors seen so far, but will look during long quantization runs